test/units/testsuite-64.sh

```diff
@@ -397,6 +414,146 @@ EOF
+    # backed by a file
+    tgtadm --lld iscsi --op new --mode target --tid=2 --targetname "$target_name"
+    # lun-0 is reserved by iSCSI
+    for i in {1..4}; do
+        tgtadm --lld iscsi --op new --mode logicalunit --tid 2 --lun "$i" -b "$mpoint/lun$i.img"
+        tgtadm --lld iscsi --op update --mode logicalunit --tid 2 --lun "$i"
+        expected_symlinks+=(
+            "/dev/disk/by-path/ip-$target_ip:$target_port-iscsi-$target_name-lun-$i"
+        )
+    done
+    tgtadm --lld iscsi --op bind --mode target --tid 2 -I ALL
+    # Configure the iSCSI initiator
+    iscsiadm --mode discoverydb --type sendtargets --portal "$target_ip" --discover
+    iscsiadm --mode node --targetname "$target_name" --portal "$target_ip:$target_port" --login
+    udevadm settle
+    # Check if all device symlinks are valid and if all expected device symlinks exist
+    for link in "${expected_symlinks[@]}"; do
+        # We need to do some active waiting anyway, as it may take kernel a bit
+        # to attach the newly connected SCSI devices
+        helper_wait_for_dev "$link"
+        test -e "$link"
+    done
+    udevadm settle
+    helper_check_device_symlinks
+    # Add all iSCSI devices into a LVM volume group, create two logical volumes,
+    # and check if necessary symlinks exist (and are valid)
+    lvm pvcreate -y "${expected_symlinks[@]}"
+    lvm pvs
+    lvm vgcreate "$vgroup" -y "${expected_symlinks[@]}"
+    lvm vgs
+    lvm vgchange -ay "$vgroup"
+    lvm lvcreate -y -L 4M "$vgroup" -n mypart1
+    lvm lvcreate -y -L 8M "$vgroup" -n mypart2
+    lvm lvs
+    udevadm settle
+    test -e "/dev/$vgroup/mypart1"
+    test -e "/dev/$vgroup/mypart2"
+    mkfs.ext4 -L mylvpart1 "/dev/$vgroup/mypart1"
+    udevadm settle
+    test -e "/dev/disk/by-label/mylvpart1"
+    helper_check_device_symlinks "/dev/disk" "/dev/$vgroup"
+    # Disconnect the iSCSI devices and check all the symlinks
+    iscsiadm --mode node --targetname "$target_name" --portal "$target_ip:$target_port" --logout
+    # "Reset" the DM state, since we yanked the backing storage from under the LVM,
+    # so the currently active VGs/LVs are invalid
+    dmsetup remove_all --deferred
+    udevadm settle
+    # The LVM and iSCSI related symlinks should be gone
+    test ! -e "/dev/$vgroup"
+    test ! -e "/dev/disk/by-label/mylvpart1"
+    for link in "${expected_symlinks[@]}"; do
+        test ! -e "$link"
+    done
+    helper_check_device_symlinks "/dev/disk"
+    # Reconnect the iSCSI devices and check if everything get detected correctly
+    iscsiadm --mode discoverydb --type sendtargets --portal "$target_ip" --discover
+    iscsiadm --mode node --targetname "$target_name" --portal "$target_ip:$target_port" --login
+    udevadm settle
+    for link in "${expected_symlinks[@]}"; do
+        helper_wait_for_dev "$link"
+        test -e "$link"
+    done
+    udevadm settle
+    # Wait for pvscan to finish
+    helper_wait_for_dev "/dev/$vgroup/mypart1"
```

Comment:
@yuwata I'm not completely sure about this part. What basically happens is that if I connect to an iSCSI target, whose LUNs contain an LVM, the LVM starts to reconstruct automatically thanks to various udev rules. However, the `/dev/<vgname>` directory may not exist even after calling `udevadm settle`, see an example log: https://gist.githubusercontent.com/mrc0mmand/682e31ff334ea99d15861d684e0d40ed/raw/e3a2df2646e9718f15638af5d7e7614b5dc9f597/iscsi-lvm-activation.txt

After looking through the relevant udev rules, I'm still a bit confused if the `/dev/<vgname` directory is actually created by udev, or udev creates just the LV symlinks in it (as stated in `11-dm-lvm.rules`). My main concern is mainly if this is an expected behavior, and the `helper_wait_for_dev()` call is justified, or if it's an actual issue.

Reply:
It seems the error is caused that pvscan is not finished yet, and no uevent about LVM volume is triggered yet. So, `udevadm settle` finishes instantaneously.
How about to add a helper which checks that corresponding pvscan@ services are once started and wait for they are finished?

BTW, why do you introduced `helper_wait_for_dev`? After `udevadm settle` finished, all devnodes or devlinks should already exist. I think we do not need to 'wait'.

Reply:
> It seems the error is caused that `pvscan` is not finished yet, and no uevent about LVM volume is triggered yet. So, `udevadm settle` finishes instantaneously.
> How about to add a helper which checks that corresponding `pvscan@` services are once started and wait for they are finished?

That sounds like a nice solution, will look into that, thanks!

> BTW, why do you introduced `helper_wait_for_dev`? After `udevadm settle` finished, all devnodes or devlinks should already exist. I think we do not need to 'wait'.

There's a window after `iscsiadm ... --login` during which the iSCSI devices get detected by kernel, so we might call `udevadm settle` before all necessary devices are properly detected (and the respective uevents are sent out).

<details>
<summary>Without the `helper_wait_for_dev()`:</summary>

```
[   10.438135] H testsuite-64.sh[595]: + iscsiadm --mode discoverydb --type sendtargets --portal 127.0.0.1 --discover
[   10.441502] H testsuite-64.sh[1256]: 127.0.0.1:3260,1 iqn.2021-09.com.example:iscsi.lvm.test
[   10.441811] H testsuite-64.sh[595]: + iscsiadm --mode node --targetname iqn.2021-09.com.example:iscsi.lvm.test --portal 127.0.0.1:3260 --login
[   10.452249] H kernel: scsi host8: iSCSI Initiator over TCP/IP
[   10.446403] H iscsid[749]: iscsid: Connection3:0 to [target: iqn.2021-09.com.example:iscsi.lvm.test, portal: 127.0.0.1,3260] through [iface: default] is operational now
[   10.454478] H kernel: scsi 8:0:0:0: RAID              IET      Controller       0001 PQ: 0 ANSI: 5
[   10.454698] H kernel: scsi 8:0:0:0: Attached scsi generic sg5 type 12
[   10.454876] H kernel: scsi 8:0:0:1: Direct-Access     IET      VIRTUAL-DISK     0001 PQ: 0 ANSI: 5
[   10.455073] H kernel: sd 8:0:0:1: Attached scsi generic sg6 type 0
[   10.455097] H kernel: sd 8:0:0:1: Power-on or device reset occurred
[   10.455276] H kernel: scsi 8:0:0:2: Direct-Access     IET      VIRTUAL-DISK     0001 PQ: 0 ANSI: 5
[   10.455535] H kernel: sd 8:0:0:1: [sdf] 65536 512-byte logical blocks: (33.6 MB/32.0 MiB)
[   10.455537] H kernel: sd 8:0:0:1: [sdf] 1024-byte physical blocks
[   10.455554] H kernel: sd 8:0:0:2: Attached scsi generic sg7 type 0
[   10.455607] H kernel: sd 8:0:0:2: Power-on or device reset occurred
[   10.455614] H kernel: sd 8:0:0:1: [sdf] Write Protect is off
[   10.455616] H kernel: sd 8:0:0:1: [sdf] Mode Sense: 69 00 10 08
[   10.455738] H kernel: scsi 8:0:0:3: Direct-Access     IET      VIRTUAL-DISK     0001 PQ: 0 ANSI: 5
[   10.455753] H kernel: sd 8:0:0:1: [sdf] Write cache: enabled, read cache: enabled, supports DPO and FUA
[   10.456061] H kernel: sd 8:0:0:3: Attached scsi generic sg8 type 0
[   10.456095] H kernel: sd 8:0:0:3: Power-on or device reset occurred
[   10.456252] H kernel: scsi 8:0:0:4: Direct-Access     IET      VIRTUAL-DISK     0001 PQ: 0 ANSI: 5
[   10.456423] H kernel: sd 8:0:0:3: [sdh] 65536 512-byte logical blocks: (33.6 MB/32.0 MiB)
[   10.456426] H kernel: sd 8:0:0:3: [sdh] 1024-byte physical blocks
[   10.456473] H kernel: sd 8:0:0:3: [sdh] Write Protect is off
[   10.456476] H kernel: sd 8:0:0:3: [sdh] Mode Sense: 69 00 10 08
[   10.456521] H kernel: sd 8:0:0:4: Attached scsi generic sg9 type 0
[   10.456674] H kernel: sd 8:0:0:4: Power-on or device reset occurred
[   10.456722] H kernel: sd 8:0:0:3: [sdh] Write cache: enabled, read cache: enabled, supports DPO and FUA
[   10.456953] H kernel: sd 8:0:0:4: [sdi] 65536 512-byte logical blocks: (33.6 MB/32.0 MiB)
[   10.456955] H kernel: sd 8:0:0:4: [sdi] 1024-byte physical blocks
[   10.457314] H kernel: sd 8:0:0:4: [sdi] Write Protect is off
[   10.457317] H kernel: sd 8:0:0:4: [sdi] Mode Sense: 69 00 10 08
[   10.458880] H kernel: sd 8:0:0:4: [sdi] Write cache: enabled, read cache: enabled, supports DPO and FUA
[   10.459146] H kernel: sd 8:0:0:2: [sdg] 65536 512-byte logical blocks: (33.6 MB/32.0 MiB)
[   10.459149] H kernel: sd 8:0:0:2: [sdg] 1024-byte physical blocks
[   10.459204] H kernel: sd 8:0:0:2: [sdg] Write Protect is off
[   10.459206] H kernel: sd 8:0:0:2: [sdg] Mode Sense: 69 00 10 08
[   10.459310] H kernel: sd 8:0:0:2: [sdg] Write cache: enabled, read cache: enabled, supports DPO and FUA
[   10.453661] H testsuite-64.sh[1257]: Logging in to [iface: default, target: iqn.2021-09.com.example:iscsi.lvm.test, portal: 127.0.0.1,3260]
[   10.453661] H testsuite-64.sh[1257]: Login to [iface: default, target: iqn.2021-09.com.example:iscsi.lvm.test, portal: 127.0.0.1,3260] successful.
[   10.454057] H testsuite-64.sh[595]: + udevadm settle
[   10.458685] H testsuite-64.sh[595]: + for link in "${expected_symlinks[@]}"
[   10.458685] H testsuite-64.sh[595]: + test -e /dev/disk/by-path/ip-127.0.0.1:3260-iscsi-iqn.2021-09.com.example:iscsi.lvm.test-lun-1
[   10.459075] H systemd[1]: Received SIGCHLD from PID 595 (bash).
[   10.459198] H systemd[1]: Child 595 (bash) died (code=exited, status=1/FAILURE)
[   10.459302] H systemd[1]: testsuite-64.service: Child 595 belongs to testsuite-64.service.
[   10.459397] H systemd[1]: testsuite-64.service: Main process exited, code=exited, status=1/FAILURE
[   10.459497] H systemd[1]: testsuite-64.service: Failed with result 'exit-code'.
[   10.459613] H audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=testsuite-64 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
[   10.459713] H systemd[1]: testsuite-64.service: Service will not restart (restart setting)
[   10.459807] H systemd[1]: testsuite-64.service: Changed start -> failed
[   10.459986] H systemd[1]: testsuite-64.service: Job 219 testsuite-64.service/start finished, result=failed
[   10.460077] H systemd[1]: Failed to start testsuite-64.service.
[   10.460481] H systemd[1]: testsuite-64.service: Unit entered failed state.
[   10.460578] H systemd[1]: testsuite-64.service: Consumed 1.177s CPU time.
[   10.460834] H systemd[1]: testsuite-64.service: Control group is empty.
[   10.460939] H systemd[1]: systemd-journald.service: Received EPOLLHUP on stored fd 54 (stored), closing.
[   10.461681] H systemd[1]: testsuite.target changed dead -> active
[   10.461841] H systemd[1]: testsuite.target: Job 139 testsuite.target/start finished, result=done
[   10.462022] H systemd[1]: Reached target testsuite.target.
[   10.462453] H sh[1267]: + systemctl poweroff --no-block
[   10.463116] H systemd[1]: Failed to read pids.max attribute of cgroup root, ignoring: No data available
[   10.463215] H systemd[1]: end.service: Passing 0 fds to service
[   10.463325] H systemd[1]: end.service: About to execute /bin/sh -x -c "systemctl poweroff --no-block"
[   10.463423] H systemd[1]: end.service: Forked /bin/sh as 1267
[   10.463909] H systemd[1]: end.service: Changed dead -> start
[   10.463972] H systemd[1]: Starting end.service...
[   10.464328] H systemd[1]: systemd-journald.service: Got notification message from PID 552 (FDSTORE=1)
[   10.464391] H systemd[1]: systemd-journald.service: Added fd 54 (n/a) to fd store.
[   10.468189] H systemd[1]: poweroff.target: Trying to enqueue job poweroff.target/start/replace-irreversibly
[   10.469377] H systemd[1]: system-serial\x2dgetty.slice: Installed new job system-serial\x2dgetty.slice/stop as 782
[   10.469465] H systemd[1]: systemd-update-done.service: Installed new job systemd-update-done.service/stop as 785
[   10.469540] H systemd[1]: systemd-tmpfiles-clean.timer: Installed new job systemd-tmpfiles-clean.timer/stop as 738
[   10.470786] H systemd[1]: systemd-poweroff.service: Installed new job systemd-poweroff.service/start as 692
[   10.471320] H systemd[1]: systemd-remount-fs.service: Installed new job systemd-remount-fs.service/stop as 760
[   10.471475] H systemd[1]: basic.target: Installed new job basic.target/stop as 720
[   10.471618] H systemd[1]: poweroff.target: Installed new job poweroff.target/start as 691
[   10.471718] H systemd[1]: paths.target: Installed new job paths.target/stop as 718
[   10.471958] H systemd[1]: slices.target: Installed new job slices.target/stop as 788
[   10.474368] H systemd[1]: swap.target: Installed new job swap.target/stop as 735
[   10.474490] H systemd[1]: systemd-coredump.socket: Installed new job systemd-coredump.socket/stop as 773
[   10.474596] H systemd[1]: final.target: Installed new job final.target/start as 695
[   10.474866] H systemd[1]: run-credentials-systemd\x2dsysusers.service.mount: Installed new job run-credentials-systemd\x2dsysusers.service.mount/stop as 697
[   10.474968] H systemd[1]: umount.target: Installed new job umount.target/start as 696
[   10.475056] H systemd[1]: multi-user.target: Installed new job multi-user.target/stop as 722
[   10.475127] H systemd[1]: local-fs.target: Installed new job local-fs.target/stop as 717
[   10.475203] H systemd[1]: testsuite.target: Installed new job testsuite.target/stop as 723
[   10.475327] H systemd[1]: user.slice: Installed new job user.slice/stop as 781
[   10.475445] H systemd[1]: systemd-user-sessions.service: Installed new job systemd-user-sessions.service/stop as 706
[   10.475542] H systemd[1]: sysinit.target: Installed new job sysinit.target/stop as 761
[   10.475690] H audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=end comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[   10.475852] H audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=end comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[   10.475933] H systemd[1]: iscsi_storeuD5.mount: Installed new job iscsi_storeuD5.mount/stop as 698
[   10.476039] H systemd[1]: getty.target: Installed new job getty.target/stop as 742
[   10.476154] H systemd[1]: systemd-ask-password-wall.path: Installed new job systemd-ask-password-wall.path/stop as 780
[   10.484034] H kernel: sd 8:0:0:1: [sdf] Attached SCSI disk
[   10.476523] H systemd[1]: dbus.socket: Installed new job dbus.socket/stop as 730
[   10.476624] H systemd[1]: shutdown.target: Installed new job shutdown.target/start as 701
[   10.476742] H systemd[1]: Successfully forked off '(sd-sync)' as PID 1271.
[   10.476874] H systemd[1]: systemd-random-seed.service: Installed new job systemd-random-seed.service/stop as 719
[   10.476990] H systemd[1]: proc-sys-fs-binfmt_misc.automount: Installed new job proc-sys-fs-binfmt_misc.automount/stop as 759
[   10.477113] H systemd[1]: end.service: Job 220 end.service/start finished, result=canceled
[   10.477394] H systemd[1]: end.service: Installed new job end.service/stop as 748
[   10.477496] H systemd[1]: systemd-logind.service: Installed new job systemd-logind.service/stop as 732
[   10.477610] H systemd[1]: local-fs-pre.target: Installed new job local-fs-pre.target/stop as 744
[   10.477716] H systemd[1]: sockets.target: Installed new job sockets.target/stop as 768
[   10.477816] H systemd[1]: tgtd.service: Installed new job tgtd.service/stop as 763
[   10.477955] H systemd[1]: system-modprobe.slice: Installed new job system-modprobe.slice/stop as 751
[   10.478060] H systemd[1]: systemd-sysusers.service: Installed new job systemd-sysusers.service/stop as 709
[   10.478200] H systemd[1]: cryptsetup.target: Installed new job cryptsetup.target/stop as 725
[   10.478311] H systemd[1]: systemd-tmpfiles-setup-dev.service: Installed new job systemd-tmpfiles-setup-dev.service/stop as 712
[   10.478408] H systemd[1]: veritysetup.target: Installed new job veritysetup.target/stop as 739
[   10.478570] H systemd[1]: ldconfig.service: Installed new job ldconfig.service/stop as 737
[   10.478678] H systemd[1]: iscsi-shutdown.service: Installed new job iscsi-shutdown.service/stop as 710
[   10.478755] H systemd[1]: systemd-sysctl.service: Installed new job systemd-sysctl.service/stop as 740
[   10.478856] H systemd[1]: iscsid.socket: Installed new job iscsid.socket/stop as 750
[   10.478926] H systemd[1]: serial-getty@ttyS0.service: Installed new job serial-getty@ttyS0.service/stop as 711
[   10.478997] H systemd[1]: systemd-update-utmp.service: Installed new job systemd-update-utmp.service/stop as 779
[   10.479091] H systemd[1]: timers.target: Installed new job timers.target/stop as 783
[   10.479172] H systemd[1]: dbus-broker.service: Installed new job dbus-broker.service/stop as 731
[   10.479256] H systemd[1]: systemd-tmpfiles-setup.service: Installed new job systemd-tmpfiles-setup.service/stop as 745
[   10.479332] H systemd[1]: system-lvm2\x2dpvscan.slice: Installed new job system-lvm2\x2dpvscan.slice/stop as 766
[   10.479465] H systemd[1]: iscsid.service: Installed new job iscsid.service/stop as 703
[   10.479532] H systemd[1]: remote-fs-pre.target: Installed new job remote-fs-pre.target/stop as 767
[   10.479616] H systemd[1]: tmp.mount: Installed new job tmp.mount/stop as 699
[   10.479835] H systemd[1]: systemd-ask-password-console.path: Installed new job systemd-ask-password-console.path/stop as 743
[   10.480013] H systemd[1]: poweroff.target: Enqueued job poweroff.target/start as 691
[   10.490061] H kernel: sd 8:0:0:4: [sdi] Attached SCSI disk
[   10.490266] H kernel: sd 8:0:0:3: [sdh] Attached SCSI disk
[   10.491426] H kernel: sd 8:0:0:2: [sdg] Attached SCSI disk
[   10.494521] H kernel: sd 8:0:0:1: [sdf] Synchronizing SCSI cache
```
</details>

Reply:
I've implemented the `helper_wait_for_service()` function for the pvscan stuff, but I'm not sure if there's any elegant solution for the other usecase of `helper_wait_for_dev`.

Reply:
Hmm, the new helper only certifies that the services are not active. But, some service may not be triggered yet...
See your original logs in gist, `pvscan` for 8:112 and 8:128 are not started when the loop below is finished.
```
    for link in "${expected_symlinks[@]}"; do
        helper_wait_for_dev "$link"
        test -e "$link"
    done
```

Reply:
> Hmm, the new helper only certifies that the services are not active. But, some service may not be triggered yet...
> See your original logs in gist, `pvscan` for 8:112 and 8:128 are not started when the loop below is finished.
> 
> ```
>     for link in "${expected_symlinks[@]}"; do
>         helper_wait_for_dev "$link"
>         test -e "$link"
>     done
> ```

Hah, good point. I guess using the concrete service names is necessary after all, I'll look into that.

Reply:
> Also, when implementing this, I kinda missed one particular functionality in `udevadm` - getting the value of a specific property (basically something like `systemctl show -P propname`) - would it make sense to implement something like that?

Sounds nice!